src/AutoRest.CSharp/MgmtTest/Generation/ScenarioTestWriter.cs

```diff
@@ -0,0 +1,347 @@
+                    {
+                        writer.Append($"Template = {typeof(System.Text.Json.JsonDocument)}.Parse({templatePayload}).RootElement");
+                        writer.Line($",");
+                    }
+
+                    var outputs = new CodeWriterDeclaration("deployOutputs");
+                    using (writer.Scope($"if ({deploymentOperation}.Value.Data.Properties.Outputs is Dictionary<string, object> {outputs:D})"))
+                    {
+                        foreach (var variableName in step.OutputVariableNames ?? new List<string>())
+                        {
+                            var element = new CodeWriterDeclaration("outputVariable");
+                            using (writer.Scope($"if ({outputs}.ContainsKey(\"{variableName}\") && {outputs}[\"{variableName}\"] is Dictionary<string, object> {element:D})"))
+                            {
+                                writer.Line($"{variableName} = {element}[\"value\"].ToString();");
+                            }
+                        }
+                    }
+                }
+            }
+            writer.Line();
+        }
+
+        protected void WriteScenarioMethod(Scenario scenario)
+        {
+            writer.Line($"[RecordedTest]");
+            using (writer.Scope($"public async Task {(scenario.ScenarioName ?? scenario.Description).ToCleanName()}()"))
+            {
+                writer.Line($"// API Scenario: {scenario.Description}");
+
+                var existedVariables = testDef.GetAllVariableNames();
+                using (new CodeWriterOavVariableScope(writer, scenario, existedVariables))
+                {
+                    WriteVariableInitializations(scenario, existedVariables);
+                    writer.Line($"resourceGroupName = Recording.GenerateAssetName(resourceGroupName);");
+                    writer.Line($"resourceGroup = (await GetArmClient().GetSubscription(new ResourceIdentifier($\"/subscriptions/{{subscriptionId}}\")).GetResourceGroups().CreateOrUpdateAsync(true, resourceGroupName, new ResourceGroupData(new AzureLocation(location)))).Value;");
+
+                    foreach (var step in scenario.Steps)
+                    {
+                        using (writer.Scope())
+                        {
+                            WriteStep(scenario, step);
+                        }
+                        writer.Line();
+                    }
+                }
+            }
+            writer.Line();
+        }
+
+        protected void WriteVariableInitializations(OavVariableScope ovs, IEnumerable<string> existedVarialbes)
+        {
+            IEnumerable<string> requiredVariables = ovs.RequiredVariables ?? new List<string>();
+            Dictionary<string, string> requiredVariableDefault = ovs.RequiredVariablesDefault ?? new Dictionary<string, string>();
```

Comment:
Only do this change when you confirm this dictionary is readonly in your code
```suggestion
            IReadOnlyDictionary<string, string> requiredVariableDefault = ovs.RequiredVariablesDefault ?? ImmutableDictionary<string, string>.Empty();
```